### PR TITLE
[Spark] Deprecate tableVersion field in type widening metadata

### PR DESCRIPTION
**Pull request by johanl-db**
## Description
The protocol specification for type widening dropped the requirement to populate a `tableVersion` field as part of type change history to track in which version of the table a given change was applied.
See protocol update: https://github.com/delta-io/delta/pull/3297
This field was used at some point during the preview but isn't needed anymore.

It is now deprecated:
- The field is preserved in table metadata that already contains it.
- The field isn't set anymore when the stable table feature is active on the table.
- The field is still set when only the preview table feature is active on the table.

The last point is necessary to avoid breaking preview clients (Delta 3.2 & Delta 4.0 preview) that require the field to be set.

## How was this patch tested?
- Updated existing metadata tests to cover `tableVersion` not being set by default.
- Added metadata tests to explicitly cover `tableVersion` being set.
- Added tests covering `tableVersion` when using the preview and stable table features.

## Does this PR introduce _any_ user-facing changes?
Yes. As of this change, a table that supports the stable table feature `typeWidening` won't have a `tableVersion` field in the type change history stored in the table metadata.

Tables that only support the preview table feature `typeWidening-preview` don't see any change.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala [hunk 1]
```diff
@@ -54,6 +56,8 @@ private[delta] case class TypeChange(
 }
 
 private[delta] object TypeChange {
+  // tableVersion was a field present during the preview and removed afterwards. We preserve it if
+  // it's already present in the type change metadata of the table to avoid breaking older clients.
```
Comment (tomvanbussel) [inline spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala:60]: What do they still use it for? Would it be possible to always set it to some constant number (e.g. zero)?

File: spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala [hunk 2]
```diff
@@ -130,11 +139,16 @@ private[delta] object TypeWideningMetadata extends DeltaLogging {
...
-        // Record the version the transaction will attempt to use in the type change metadata. If
-        // there's a conflict with another transaction, the version in the metadata will be updated
-        // during conflict resolution. See [[ConflictChecker.updateTypeWideningMetadata()]].
-        val typeChanges =
-          collectTypeChanges(oldField.dataType, newField.dataType, txn.getFirstAttemptVersion)
+        var typeChanges = collectTypeChanges(oldField.dataType, newField.dataType)
+        // The version field isn't used anymore but we need to populate it in case the table doesn't
+        // use the stable feature, as preview clients may then still access the table and rely on
+        // the field being present.
+        if (!txn.protocol.isFeatureSupported(TypeWideningTableFeature)) {
```
Comment (tomvanbussel) [inline spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala:146]: Why do we check for `TypeWideningTableFeature` not being there instead of `TypeWideningPreviewTableFeature` being there?

File: spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningMetadataSuite.scala [hunk 3]
```diff
@@ -49,8 +50,9 @@ trait TypeWideningMetadataTests extends QueryTest with DeltaSQLCommandTest {
 
   override protected def beforeAll(): Unit = {
     super.beforeAll()
-    sql(s"CREATE TABLE $testTableName (a int) USING delta " +
-      s"TBLPROPERTIES ('${DeltaConfigs.ENABLE_TYPE_WIDENING.key}' = 'true')")
+    sql(s"CREATE TABLE $testTableName (a int) USING delta TBLPROPERTIES (" +
+      s"'${DeltaConfigs.ENABLE_TYPE_WIDENING.key}' = 'true', " +
+      s"'${propertyKey(TypeWideningTableFeature)}' = 'supported')")
```
Comment (tomvanbussel) [inline spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningMetadataSuite.scala:55]: Why is this necessary? Isn't the feature automatically enabled when setting the property?

**Review by tomvanbussel — APPROVED** [review]
Just a few comments, LGTM otherwise

File: spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningMetadataSuite.scala (hunk 3 above)
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningMetadataSuite.scala:58]: The *preview* feature is enabled by default, this forces the stable to be added instead.

I added a comment to clarify that and that we're testing with stable feature by default and only explicit set the preview feature in specific tests that cover the old behavior re: `tableVersion`.
+ added a missing test for the preview feature

**Review by bart-samwel — APPROVED** [review]
Looks good to me!
